### DISCUSSION TITLE
Learn GO to simplify the cross compilation

### DISCUSSION THREAD

**Discussion by taozuhong** [mesonbuild/meson#14039] [Ideas]
I touched GO language recently, found its cross-compilation support is very simple, so cloud Meson learn its way to simplify the cross-compilation support?

GO language has two variables to control the cross-compilation: OS / ARCH, 
maybe Meson could support four control flags：
```
OS: The operation system: Harmony / Windows / Linux / macOS / iOS / Android / etc
ARCH: CPU architect: X86 / AARCH / RISCV / LONGARCH / MIPS / etc
SIZE: Memory size: 32 / 64 / 96 / 128          ---> May removed for 64 bits only
MIX: Mixed method to keep code be safe: ？？？ ---> May removed for static language
```

**Comment by eli-schwartz** [comment]
There is an open feature request asking for golang support: https://github.com/mesonbuild/meson/issues/123

Someone will have to volunteer to implement it though.

**Reply by taozuhong** [reply]
It seems very simple, just collect and build a map for cross compilation, then check and hint user to install the libs, or build it directly.

**Reply by eli-schwartz** [reply]
You can simplify any task by pointing out that it is "very simple: just do it!"

Practical implementation is harder. I can't voice an opinion on the topic until someone submits a PR containing working code to do it.

**Reply by taozuhong** [reply]
Sorry for making you feel bad, 

Do you think that it's the right way to implement it?

**Reply by eli-schwartz** [reply]
> Do you think that it's the right way to implement it?

The first step is probably to update https://github.com/mesonbuild/meson/tree/master/mesonbuild/compilers

In detect.py, we'd need to search and find out if the `go` program is installed. In (new file) go.py, we'd need to define a subclass of compilers.py -> the Compiler object, which implements various generic interfaces such as sanity check (compile a minimal hello world program) and output_args, depfile handling, a mechanism for translating C compiler flags when using CGO, etc.

In mesonbuild/backend/ we'd need to e.g. write ninja rules for `go build`, validate and error out if a single target contains both golang and non-golang sources...

I know roughly what it takes to add a new language to meson. I don't know anything about Go, so I truly can't say how easy implementing it would be as I don't have experience with the go compiler. :)

**Reply by taozuhong** [reply]
Oh, seems that you yet not got me, I want to simplify Meson's cross compilation support for existed languages, not for Go language.
